### PR TITLE
[composite compliance] batch_norm

### LINKED ISSUE #76283: [ubsan] Undefined behaviour when a running forwardAD CompositeCompliance on `batch_norm`
### 🐛 Describe the bug

PR adding the test: https://github.com/pytorch/pytorch/pull/75178
```
test_forward_ad_nn_functional_batch_norm_cpu_float32 (__main__.TestCompositeComplianceCPU) ... /var/lib/jenkins/workspace/aten/src/ATen/native/Normalization.cpp:191:31: runtime error: -5.368e+38 is outside the range of representable values of type 'float'
[6456](https://github.com/pytorch/pytorch/runs/6134744920?check_suite_focus=true#step:9:6456)
    #0 0x7ff12d00b5c2 in std::tuple<at::Tensor, at::Tensor> at::native::batch_norm_cpu_update_stats_template<float, at::native::InvStd>(at::Tensor const&, at::Tensor const&, at::Tensor const&, double, double)::'lambda'(long, long)::operator()(long, long) const (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0xc0d35c2)
[6457](https://github.com/pytorch/pytorch/runs/6134744920?check_suite_focus=true#step:9:6457)
    #1 0x7ff12cfe75c9 in .omp_outlined..175 (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0xc0af5c9)
[6458](https://github.com/pytorch/pytorch/runs/6134744920?check_suite_focus=true#step:9:6458)
    #2 0x7ff11829fbb2 in __kmp_invoke_microtask (/opt/conda/lib/libiomp5.so+0x13fbb2)
[6459](https://github.com/pytorch/pytorch/runs/6134744920?check_suite_focus=true#step:9:6459)
    #3 0x7ff11821b902 in __kmp_invoke_task_func (/opt/conda/lib/libiomp5.so+0xbb902)
[6460](https://github.com/pytorch/pytorch/runs/6134744920?check_suite_focus=true#step:9:6460)
    #4 0x7ff11821d3dd in __kmp_fork_call (/opt/conda/lib/libiomp5.so+0xbd3dd)
[6461](https://github.com/pytorch/pytorch/runs/6134744920?check_suite_focus=true#step:9:6461)
    #5 0x7ff1181ddcb4 in __kmpc_fork_call (/opt/conda/lib/libiomp5.so+0x7dcb4)
[6462](https://github.com/pytorch/pytorch/runs/6134744920?check_suite_focus=true#step:9:6462)
    #6 0x7ff12d00b825 in void at::internal::invoke_parallel<std::tuple<at::Tensor, at::Tensor> at::native::batch_norm_cpu_update_stats_template<float, at::native::InvStd>(at::Tensor const&, at::Tensor const&, at::Tensor const&, double, double)::'lambda'(long, long)>(long, long, long, std::tuple<at::Tensor, at::Tensor> at::native::batch_norm_cpu_update_stats_template<float, at::native::InvStd>(at::Tensor const&, at::Tensor const&, at::Tensor const&, double, double)::'lambda'(long, long) const&) (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0xc0d3825)
```

Ref: https://github.com/pytorch/pytorch/runs/6134744920?check_suite_focus=true#step:9:6455

cc: @zou3519 

### Versions

master

cc @ezyang @gchanan @zou3519 @Chillee @albanD @samdow

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/76283
Ref: #69991

### CODE REVIEW

**Comment by kshitij12345** [comment]
Ping @zou3519

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -188,18 +188,19 @@ std::tuple<Tensor,Tensor> batch_norm_cpu_update_stats_template(
     auto _var_sum_a = _var_sum.accessor<param_t, 1>();
 
     batch_norm_cpu_collect_stats_stub(kCPU, _mean, _var_sum, input);
-
+    auto momentum_ = static_cast<param_t>(momentum);
```
Comment (zou3519) [inline aten/src/ATen/native/Normalization.cpp:191]: Why is the static cast necessary? (what was the composite compliance problem?)

File: aten/src/ATen/native/Normalization.cpp [hunk 2]
```diff
@@ -188,18 +188,19 @@ std::tuple<Tensor,Tensor> batch_norm_cpu_update_stats_template(
...
     parallel_for(0, n_input, 1, [&](int64_t b_begin, int64_t b_end) {
       for (const auto f : c10::irange(b_begin, b_end)) {
         save_mean_a[f] = _mean_a[f];
         save_var_transform_a[f] = VarTransform<accscalar_t>{}(_var_sum_a[f] / n, eps);
 
         if (running_mean.defined()) {
-          running_mean_a[f] = momentum * _mean_a[f] + (1 - momentum) * running_mean_a[f];
+          running_mean_a[f] = momentum_ * _mean_a[f] + (1 - momentum_) * running_mean_a[f];
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply aten/src/ATen/native/Normalization.cpp:198]: Yes. It does. Makes sense.

**Review by zou3519 — APPROVED** [review]
LGTM as long as tests pass

File: aten/src/ATen/native/Normalization.cpp (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/native/Normalization.cpp]: This won't work as the computation in `double` results in values out of `float`'s range.

```c++
int main() {
    double d = -5.368e+38;
    float f = static_cast<float>(d);
}
```

Following code compiled as `clang++ test_ubsan.cc -fsanitize=undefined`
Fails with
```
test_ubsan.cc:3:15: runtime error: -5.368e+38 is outside the range of representable values of type 'float'
    #0 0x4242f1  (/home/kshiteej/Pytorch/pytorch_functorch/test/a.out+0x4242f1)
    #1 0x7fef65587d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
    #2 0x7fef65587e3f  (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f)
```

File: aten/src/ATen/native/Normalization.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kshitij12345) [reply aten/src/ATen/native/Normalization.cpp]: That didn't work : https://github.com/pytorch/pytorch/pull/79990#discussion_r948765491
